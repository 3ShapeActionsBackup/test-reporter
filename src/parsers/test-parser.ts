import {Endpoints} from '@octokit/types'

type OutputParameters = Endpoints['POST /repos/:owner/:repo/check-runs']['parameters']['output']

export type ParseTestResult = (content: string) => Promise<TestResult>

export interface TestResult {
  success: boolean
  output: OutputParameters
}
