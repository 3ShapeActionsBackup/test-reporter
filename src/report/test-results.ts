export class TestRunResult {
  constructor(readonly suites: TestSuiteResult[], private totalTime?: number) {}

  get tests(): number {
    return this.suites.reduce((sum, g) => sum + g.tests, 0)
  }

  get passed(): number {
    return this.suites.reduce((sum, g) => sum + g.passed, 0)
  }
  get failed(): number {
    return this.suites.reduce((sum, g) => sum + g.failed, 0)
  }
  get skipped(): number {
    return this.suites.reduce((sum, g) => sum + g.skipped, 0)
  }

  get time(): number {
    return this.totalTime ?? this.suites.reduce((sum, g) => sum + g.time, 0)
  }

  get result(): TestExecutionResult {
    return this.suites.some(t => t.result === 'failed') ? 'failed' : 'success'
  }
}

export class TestSuiteResult {
  constructor(readonly name: string, readonly groups: TestGroupResult[], private totalTime?: number) {}

  get tests(): number {
    return this.groups.reduce((sum, g) => sum + g.tests.length, 0)
  }

  get passed(): number {
    return this.groups.reduce((sum, g) => sum + g.passed, 0)
  }
  get failed(): number {
    return this.groups.reduce((sum, g) => sum + g.failed, 0)
  }
  get skipped(): number {
    return this.groups.reduce((sum, g) => sum + g.skipped, 0)
  }
  get time(): number {
    return this.totalTime ?? this.groups.reduce((sum, g) => sum + g.time, 0)
  }

  get result(): TestExecutionResult {
    return this.groups.some(t => t.result === 'failed') ? 'failed' : 'success'
  }
}

export class TestGroupResult {
  constructor(readonly name: string | undefined | null, readonly tests: TestCaseResult[]) {}

  get passed(): number {
    return this.tests.reduce((sum, t) => (t.result === 'success' ? sum + 1 : sum), 0)
  }
  get failed(): number {
    return this.tests.reduce((sum, t) => (t.result === 'failed' ? sum + 1 : sum), 0)
  }
  get skipped(): number {
    return this.tests.reduce((sum, t) => (t.result === 'skipped' ? sum + 1 : sum), 0)
  }
  get time(): number {
    return this.tests.reduce((sum, t) => sum + t.time, 0)
  }

  get result(): TestExecutionResult {
    return this.tests.some(t => t.result === 'failed') ? 'failed' : 'success'
  }
}

export class TestCaseResult {
  constructor(readonly name: string, readonly result: TestExecutionResult, readonly time: number) {}
}

export type TestExecutionResult = 'success' | 'skipped' | 'failed' | undefined
