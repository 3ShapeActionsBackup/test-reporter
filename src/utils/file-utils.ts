import * as fs from 'fs'

export function getFileContent(path: string): string {
  if (!fs.existsSync(path)) {
    throw new Error(`File '${path}' not found`)
  }

  if (!fs.lstatSync(path).isFile()) {
    throw new Error(`'${path}' is not a file`)
  }

  return fs.readFileSync(path, {encoding: 'utf8'})
}

export function normalizeDirPath(path: string, trailingSeparator: boolean): string {
  if (!path) {
    return path
  }

  path = normalizeFilePath(path)
  if (trailingSeparator && !path.endsWith('/')) {
    path += '/'
  }
  return path
}

export function normalizeFilePath(path: string): string {
  if (!path) {
    return path
  }

  return path.trim().replace(/\\/g, '/')
}
