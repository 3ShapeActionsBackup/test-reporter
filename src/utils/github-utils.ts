import * as github from '@actions/github'
import {EventPayloads} from '@octokit/webhooks'

export function getCheckRunSha(): string {
  if (github.context.payload.pull_request) {
    const pr = github.context.payload.pull_request as EventPayloads.WebhookPayloadPullRequestPullRequest
    return pr.head.sha
  }

  return github.context.sha
}
