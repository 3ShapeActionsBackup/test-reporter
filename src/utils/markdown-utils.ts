export enum Align {
  Left = ':---',
  Center = ':---:',
  Right = '---:',
  None = '---'
}

export const Icon = {
  skip: '✖️', // ':heavy_multiplication_x:'
  success: '✔️', // ':heavy_check_mark:'
  fail: '❌' // ':x:'
}

export function details(summary: string, content: string): string {
  return `<details><summary>${summary}</summary>${content}</details>`
}

export function link(title: string, address: string): string {
  return `[${title}](${address})`
}

type ToString = string | number | boolean | Date
export function table(headers: ToString[], align: ToString[], ...rows: ToString[][]): string {
  const headerRow = `| ${headers.map(tableEscape).join(' | ')} |`
  const alignRow = `| ${align.join(' | ')} |`
  const contentRows = rows.map(row => `| ${row.map(tableEscape).join(' | ')} |`).join('\n')
  return [headerRow, alignRow, contentRows].join('\n')
}

export function tableEscape(content: ToString): string {
  return content.toString().replace('|', '\\|')
}
