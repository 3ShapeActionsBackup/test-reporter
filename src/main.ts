import * as core from '@actions/core'
import * as github from '@actions/github'
import {parseJestJunit} from './parsers/jest-junit/jest-junit-parser'
import {ParseTestResult} from './parsers/test-parser'
import {getFileContent} from './utils/file-utils'
import {getCheckRunSha} from './utils/github-utils'

async function run(): Promise<void> {
  try {
    await main()
  } catch (error) {
    core.setFailed(error.message)
  }
}

async function main(): Promise<void> {
  const name = core.getInput('name', {required: true})
  const path = core.getInput('path', {required: true})
  const reporter = core.getInput('reporter', {required: true})
  const token = core.getInput('token', {required: true})

  const octokit = github.getOctokit(token)
  const sha = getCheckRunSha()

  const parser = getParser(reporter)
  const content = getFileContent(path)
  const result = await parser(content)

  await octokit.checks.create({
    head_sha: sha,
    name,
    status: 'completed',
    conclusion: result.success ? 'success' : 'failure',
    output: result.output,
    ...github.context.repo
  })
}

function getParser(reporter: string): ParseTestResult {
  switch (reporter) {
    case 'dotnet-trx':
      throw new Error('Not implemented yet!')
    case 'flutter-machine':
      throw new Error('Not implemented yet!')
    case 'jest-junit':
      return parseJestJunit
    default:
      throw new Error(`Input parameter 'reporter' is set to invalid value '${reporter}'`)
  }
}

run()
