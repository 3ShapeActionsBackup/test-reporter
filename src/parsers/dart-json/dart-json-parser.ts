import {ParseOptions, TestResult} from '../test-parser'

import {Icon} from '../../utils/markdown-utils'

import {
  ReportEvent,
  Suite,
  Group,
  TestStartEvent,
  TestDoneEvent,
  ErrorEvent,
  isSuiteEvent,
  isGroupEvent,
  isTestStartEvent,
  isTestDoneEvent,
  isErrorEvent,
  isDoneEvent
} from './dart-json-types'

class TestRun {
  constructor(readonly suites: TestSuite[], readonly success: boolean, readonly time: number) {}
  get count(): number {
    return Object.values(this.suites).reduce((sum, g) => sum + g.count, 0)
  }
  get passed(): number {
    return Object.values(this.suites).reduce((sum, g) => sum + g.passed, 0)
  }
  get failed(): number {
    return Object.values(this.suites).reduce((sum, g) => sum + g.failed, 0)
  }
  get skipped(): number {
    return Object.values(this.suites).reduce((sum, g) => sum + g.skipped, 0)
  }
}

class TestSuite {
  constructor(readonly suite: Suite) {}
  tests = new TestGroup()
  groups: {[id: number]: TestGroup} = {}
  get count(): number {
    return this.tests.count + Object.values(this.groups).reduce((sum, g) => sum + g.count, 0)
  }
  get passed(): number {
    return this.tests.passed + Object.values(this.groups).reduce((sum, g) => sum + g.passed, 0)
  }
  get failed(): number {
    return this.tests.failed + Object.values(this.groups).reduce((sum, g) => sum + g.failed, 0)
  }
  get skipped(): number {
    return this.tests.skipped + Object.values(this.groups).reduce((sum, g) => sum + g.skipped, 0)
  }
  get time(): number {
    return this.tests.time + Object.values(this.groups).reduce((sum, g) => sum + g.time, 0)
  }
}

class TestGroup {
  constructor(readonly group?: Group) {}
  tests: TestSuiteTest[] = []
  get count(): number {
    return this.tests.length
  }
  get passed(): number {
    return this.tests.reduce((sum, t) => (t.isPassed ? sum + 1 : sum), 0)
  }
  get failed(): number {
    return this.tests.reduce((sum, t) => (t.isFailed ? sum + 1 : sum), 0)
  }
  get skipped(): number {
    return this.tests.reduce((sum, t) => (t.isSkipped ? sum + 1 : sum), 0)
  }
  get time(): number {
    return this.tests.reduce((sum, t) => sum + t.time, 0)
  }
}

class TestSuiteTest {
  constructor(readonly testStart: TestStartEvent) {
    this.groupId = testStart.test.groupIDs[testStart.test.groupIDs.length - 1]
  }
  readonly groupId: number
  testDone?: TestDoneEvent
  error?: ErrorEvent
  get isPassed(): boolean {
    return this.testDone?.result === 'success' && !this.testDone?.skipped
  }
  get isFailed(): boolean {
    return this.testDone?.result !== 'success'
  }
  get isSkipped(): boolean {
    return this.testDone?.skipped === true
  }
  get time(): number {
    return this.testDone !== undefined ? this.testDone.time - this.testStart.time : 0
  }
}

export async function parseDartJson(content: string, options: ParseOptions): Promise<TestResult> {
  const testRun = getTestRun(content)
  const icon = testRun.success ? Icon.success : Icon.fail

  return {
    success: testRun.success,
    output: {
      title: `${options.name.trim()} ${icon}`,
      summary: getSummary(testRun),
      annotations: options.annotations ? [] : undefined
    }
  }
}

function getTestRun(content: string): TestRun {
  const lines = content.split(/\n\r?/g).filter(line => line !== '')
  const events = lines.map(str => JSON.parse(str)) as ReportEvent[]

  let success = false
  let totalTime = 0
  const suites: {[id: number]: TestSuite} = {}
  const tests: {[id: number]: TestSuiteTest} = {}

  for (const evt of events) {
    if (isSuiteEvent(evt)) {
      suites[evt.suite.id] = new TestSuite(evt.suite)
    } else if (isGroupEvent(evt)) {
      suites[evt.group.suiteID].groups[evt.group.id] = new TestGroup(evt.group)
    } else if (isTestStartEvent(evt) && evt.test.url !== null) {
      const test: TestSuiteTest = new TestSuiteTest(evt)
      const suite = suites[evt.test.suiteID]
      const group =
        evt.test.groupIDs.length === 0 ? suite.tests : suite.groups[evt.test.groupIDs[evt.test.groupIDs.length - 1]]
      group.tests.push(test)
      tests[evt.test.id] = test
    } else if (isTestDoneEvent(evt) && !evt.hidden) {
      tests[evt.testID].testDone = evt
    } else if (isErrorEvent(evt)) {
      tests[evt.testID].error = evt
    } else if (isDoneEvent(evt)) {
      success = evt.success
      totalTime = evt.time
    }
  }

  return new TestRun(Object.values(suites), success, totalTime)
}

function getSummary(testRun: TestRun): string {
  const tests = testRun.count
  const time = `${(testRun.time / 1000).toFixed(3)}s`
  const passed = testRun.passed
  const skipped = testRun.skipped
  const failed = testRun.failed

  const headingLine = `**${tests}** tests were completed in **${time}** with **${passed}** passed, **${skipped}** skipped and **${failed}** failed.`
  return `${headingLine}`
}
