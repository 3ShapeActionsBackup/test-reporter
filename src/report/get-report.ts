import * as core from '@actions/core'
import {TestExecutionResult, TestRunResult, TestSuiteResult} from './test-results'
import {Align, formatTime, Icon, link, table} from '../utils/markdown-utils'
import {slug} from '../utils/slugger'

export interface ReportOptions {
  listSuites?: 'all' | 'failed'
  listTests?: 'all' | 'failed' | 'none'
}

export function getReport(results: TestRunResult[], options: ReportOptions = {}): string {
  core.info('Generating check run summary')

  const maxReportLength = 65535
  const sections: string[] = []

  applySort(results)

  const badge = getReportBadge(results)
  sections.push(badge)

  const runs = getTestRunsReport(results, options)
  sections.push(...runs)

  const report = sections.join('\n')
  if (report.length > maxReportLength) {
    let msg = `**Check Run summary limit of ${maxReportLength} chars was exceed**`
    if (options.listTests !== 'all') {
      msg += '\n- Consider setting `list-tests` option to `only-failed` or `none`'
    }
    if (options.listSuites !== 'all') {
      msg += '\n- Consider setting `list-suites` option to `only-failed`'
    }

    return `${badge}\n${msg}`
  }

  return report
}

function applySort(results: TestRunResult[]): void {
  results.sort((a, b) => a.path.localeCompare(b.path))
  for (const res of results) {
    res.suites.sort((a, b) => a.name.localeCompare(b.name))
  }
}

function getReportBadge(results: TestRunResult[]): string {
  const passed = results.reduce((sum, tr) => sum + tr.passed, 0)
  const skipped = results.reduce((sum, tr) => sum + tr.skipped, 0)
  const failed = results.reduce((sum, tr) => sum + tr.failed, 0)
  return getBadge(passed, failed, skipped)
}

function getBadge(passed: number, failed: number, skipped: number): string {
  const text = []
  if (passed > 0) { text.push(`${passed} passed`) }
  if (failed > 0) { text.push(`${failed} failed`) }
  if (skipped > 0) { text.push(`${skipped} skipped`) }
  let message = text.length > 0 ? text.join(', ') : 'none'

  let color = 'success'
  if (failed > 0) {
    color = 'critical'
  } else if (passed === 0 && failed === 0) {
    color = 'yellow'
  }
  const hint = failed > 0 ? 'Tests failed' : 'Tests passed successfully'
  const uri = encodeURIComponent(`tests-${message}-${color}`)
  return `![${hint}](https://img.shields.io/badge/${uri})`
}

function getTestRunsReport(testRuns: TestRunResult[], options: ReportOptions): string[] {
  const sections: string[] = []

  if (testRuns.length > 1) {
    const tableData = testRuns.map((tr, runIndex) => {
      const time = formatTime(tr.time)
      const name = tr.path
      const addr = makeRunSlug(runIndex).link
      const nameLink = link(name, addr)
      const passed = tr.passed > 0 ? `${tr.passed}${Icon.success}` : ''
      const failed = tr.failed > 0 ? `${tr.failed}${Icon.fail}` : ''
      const skipped = tr.skipped > 0 ? `${tr.skipped}${Icon.skip}` : ''
      return [nameLink, passed, failed, skipped, time]
    })

    const resultsTable = table(
      ['Report', 'Passed', 'Failed', 'Skipped', 'Time'],
      [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right],
      ...tableData
    )
    sections.push(resultsTable)
  }

  const suitesReports = testRuns.map((tr, i) => getSuitesReport(tr, i, options)).flat()
  sections.push(...suitesReports)
  return sections
}

function getSuitesReport(tr: TestRunResult, runIndex: number, options: ReportOptions): string[] {
  const sections: string[] = []

  const time = `${(tr.time / 1000).toFixed(3)}s`

  const slug = makeRunSlug(runIndex)
  const nameLink = `<a id="${slug.id}" href="${slug.link}">${tr.path}</a>`
  const icon = getResultIcon(tr.result)
  sections.push(`## ${nameLink} ${icon}`)

  const headingLine2 = `**${tr.tests}** tests were completed in **${time}** with **${tr.passed}** passed, **${tr.failed}** failed and **${tr.skipped}** skipped.`
  sections.push(headingLine2)

  const suites = options.listSuites === 'failed' ? tr.failedSuites : tr.suites
  if (suites.length > 0) {
    const suitesTable = table(
      ['Test suite', 'Passed', 'Failed', 'Skipped', 'Time'],
      [Align.Left, Align.Right, Align.Right, Align.Right, Align.Right],
      ...suites.map((s, suiteIndex) => {
        const tsTime = formatTime(s.time)
        const tsName = s.name
        const skipLink = options.listTests === 'none' || (options.listTests === 'failed' && s.result !== 'failed')
        const tsAddr = makeSuiteSlug(runIndex, suiteIndex).link
        const tsNameLink = skipLink ? tsName : link(tsName, tsAddr)
        const passed = s.passed > 0 ? `${s.passed}${Icon.success}` : ''
        const failed = s.failed > 0 ? `${s.failed}${Icon.fail}` : ''
        const skipped = s.skipped > 0 ? `${s.skipped}${Icon.skip}` : ''
        return [tsNameLink, passed, failed, skipped, tsTime]
      })
    )
    sections.push(suitesTable)
  }

  if (options.listTests !== 'none') {
    const tests = suites
      .map((ts, suiteIndex) => getTestsReport(ts, runIndex, suiteIndex, options))
      .flat()

    if (tests.length > 1) {
      sections.push(...tests)
    }
  }

  return sections
}

function getTestsReport(ts: TestSuiteResult, runIndex: number, suiteIndex: number, options: ReportOptions): string[] {
  const groups = options.listTests === 'failed' ? ts.failedGroups : ts.groups
  if (groups.length === 0) {
    return []
  }

  const sections: string[] = []

  const tsName = ts.name
  const tsSlug = makeSuiteSlug(runIndex, suiteIndex)
  const tsNameLink = `<a id="${tsSlug.id}" href="${tsSlug.link}">${tsName}</a>`
  const icon = getResultIcon(ts.result)
  sections.push(`### ${tsNameLink} ${icon}`)

  const headingLine2 = `**${ts.tests}** tests were completed in **${ts.time}ms** with **${ts.passed}** passed, **${ts.failed}** failed and **${ts.skipped}** skipped.`
  sections.push(headingLine2)

  for (const grp of groups) {
    const tests = options.listTests === 'failed' ? grp.failedTests : grp.tests
    if (tests.length === 0) {
      continue
    }
    const grpHeader = grp.name ? `\n**${grp.name}**` : ''
    const testsTable = table(
      ['Result', 'Test', 'Time'],
      [Align.Center, Align.Left, Align.Right],
      ...grp.tests.map(tc => {
        const name = tc.name
        const time = formatTime(tc.time)
        const result = getResultIcon(tc.result)
        return [result, name, time]
      })
    )

    sections.push(grpHeader, testsTable)
  }

  return sections
}

function makeRunSlug(runIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}`)
}

function makeSuiteSlug(runIndex: number, suiteIndex: number): {id: string; link: string} {
  // use prefix to avoid slug conflicts after escaping the paths
  return slug(`r${runIndex}s${suiteIndex}`)
}

function getResultIcon(result: TestExecutionResult): string {
  switch (result) {
    case 'success':
      return Icon.success
    case 'skipped':
      return Icon.skip
    case 'failed':
      return Icon.fail
    default:
      return ''
  }
}
