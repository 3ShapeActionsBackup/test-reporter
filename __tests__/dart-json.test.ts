import * as fs from 'fs'
import * as path from 'path'

import {DartJsonParser} from '../src/parsers/dart-json/dart-json-parser'
import {ParseOptions} from '../src/test-parser'
import {getReport} from '../src/report/get-report'
import {normalizeFilePath} from '../src/utils/file-utils'

describe('dart-json tests', () => {
  it('matches report snapshot', async () => {
    const opts: ParseOptions = {
      parseErrors: true,
      trackedFiles: ['lib/main.dart', 'test/main_test.dart', 'test/second_test.dart'],
      workDir: 'C:/Users/Michal/Workspace/dorny/test-check/reports/dart/'
    }

    const fixturePath = path.join(__dirname, 'fixtures', 'dart-json.json')
    const outputPath = path.join(__dirname, '__outputs__', 'dart-json.md')
    const filePath = normalizeFilePath(path.relative(__dirname, fixturePath))
    const fileContent = fs.readFileSync(fixturePath, {encoding: 'utf8'})

    const parser = new DartJsonParser(opts)
    const result = await parser.parse(filePath, fileContent)
    expect(result).toMatchSnapshot()

    const report = getReport([result])
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, report)
  })
})
