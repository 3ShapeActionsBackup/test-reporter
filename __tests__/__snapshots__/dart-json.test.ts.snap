// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dart-json tests matches report snapshot 1`] = `
Object {
  "annotations": Array [
    Object {
      "annotation_level": "failure",
      "end_line": 13,
      "message": "Expected: <2>
  Actual: <1>


package:test_api          expect
test\\\\main_test.dart 13:9  main.<fn>.<fn>.<fn>
",
      "path": "test/main_test.dart",
      "start_line": 13,
      "title": "[test\\\\main_test.dart] Test 1 Test 1.1 Failing test",
    },
    Object {
      "annotation_level": "failure",
      "end_line": 2,
      "message": "Exception: Some error

package:darttest/main.dart 2:3  throwError
test\\\\main_test.dart 17:9        main.<fn>.<fn>.<fn>
",
      "path": "lib/main.dart",
      "start_line": 2,
      "title": "[test\\\\main_test.dart] Test 1 Test 1.1 Exception in target unit",
    },
    Object {
      "annotation_level": "failure",
      "end_line": 24,
      "message": "Exception: Some error

test\\\\main_test.dart 24:7  main.<fn>.<fn>
",
      "path": "test/main_test.dart",
      "start_line": 24,
      "title": "[test\\\\main_test.dart] Test 2 Exception in test",
    },
    Object {
      "annotation_level": "failure",
      "end_line": 5,
      "message": "TimeoutException after 0:00:00.000001: Test timed out after 0 seconds.

dart:isolate  _RawReceivePortImpl._handleMessage
",
      "path": "test/second_test.dart",
      "start_line": 5,
      "title": "[test\\\\second_test.dart] Timeout test",
    },
  ],
  "summary": "### fixtures/dart-json.json

**6** tests were completed in **3.760s** with **1** passed, **1** skipped and **4** failed.

| Result | Suite | Tests | Time | Passed ✔️ | Skipped ✖️ | Failed ❌ |
| :---: | :--- | ---: | ---: | ---: | ---: | ---: |
| ❌ | [test\\\\main_test.dart](#ts-0-test-maintest-dart) | 4 | 74ms | 1 | 0 | 3 |
| ❌ | [test\\\\second_test.dart](#ts-1-test-secondtest-dart) | 2 | 51ms | 0 | 1 | 1 |

# Test Suites

## <a id=\\"user-content-ts-0-test-maintest-dart\\" href=\\"#ts-0-test-maintest-dart\\">test\\\\main_test.dart</a> ❌

### Test 1

| Result | Test | Time |
| :---: | :--- | ---: |
| ✔️ | Test 1 Passing test | 36ms |

### Test 1 Test 1.1

| Result | Test | Time |
| :---: | :--- | ---: |
| ❌ | Test 1 Test 1.1 Failing test | 20ms |
| ❌ | Test 1 Test 1.1 Exception in target unit | 6ms |

### Test 2

| Result | Test | Time |
| :---: | :--- | ---: |
| ❌ | Test 2 Exception in test | 12ms |

## <a id=\\"user-content-ts-1-test-secondtest-dart\\" href=\\"#ts-1-test-secondtest-dart\\">test\\\\second_test.dart</a> ❌

| Result | Test | Time |
| :---: | :--- | ---: |
| ❌ | Timeout test | 37ms |
| ✖️ | Skipped test | 14ms |
",
  "title": "Dart tests ❌",
}
`;
